orc8r/cloud/go/swagger/combine.go

```diff
@@ -0,0 +1,197 @@
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package swagger
+
+import (
+	"regexp"
+	"sort"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// Combine multiple Swagger specs, giving precedence to the "common" spec.
+// First error contains merge warnings for overwritten fields and
+// incompatible Swagger specs which will be explicitly ignored in the merging
+// process.
+// This custom-built functionality mirrors the "official" implementation.
+// See: https://github.com/go-openapi/analysis/blob/master/mixin.go
+func Combine(common string, spec []string) (string, error, error) {
+	warnings := &multierror.Error{}
+	specs, commonSpec, errs := unmarshalToSwagger(common, spec)
+	if errs != nil {
+		multierror.Append(warnings, errs)
+	}
+
+	combinedSpec, errs := combine(commonSpec, specs)
+	if errs != nil {
+		multierror.Append(warnings, errs)
+	}
+
+	out, err := marshalToYAML(combinedSpec)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return out, warnings, nil
+}
+
+func combine(common SwaggerSpec, specs []SwaggerSpec) (SwaggerSpec, error) {
+	var tags [][]TagDefinition
+	var paths []map[string]interface{}
+	var responses []map[string]interface{}
+	var parameters []map[string]interface{}
+	var definitions []map[string]interface{}
+
+	for _, s := range specs {
+		tags = append(tags, s.Tags)
+		paths = append(paths, s.Paths)
+		responses = append(responses, s.Responses)
+		parameters = append(parameters, s.Parameters)
+		definitions = append(definitions, s.Definitions)
+	}
+
+	errs := &multierror.Error{}
+
+	out := SwaggerSpec{
+		Swagger:  common.Swagger,
+		Info:     common.Info,
+		BasePath: common.BasePath,
+		Consumes: common.Consumes,
+		Produces: common.Produces,
+		Schemes:  common.Schemes,
+
+		Tags:        combineTags(common.Tags, tags, errs),
+		Paths:       combineSubSpec(common.Paths, paths, "paths", errs),
+		Responses:   combineSubSpec(common.Responses, responses, "responses", errs),
+		Parameters:  combineSubSpec(common.Parameters, parameters, "parameters", errs),
+		Definitions: combineSubSpec(common.Definitions, definitions, "definitions", errs),
+	}
+
+	return out, errs.ErrorOrNil()
+}
+
+// unmarshalToSwagger converts a list of specs and a common spec
+// from YAML format to Swagger structs.
+// Returned Error is a list of errors from incompatible swagger specs
+func unmarshalToSwagger(yamlCommon string, yamlSpecs []string) ([]SwaggerSpec, SwaggerSpec, error) {
+	errs := &multierror.Error{}
+
+	editedYAMLSpecs := makeAllYAMLReferencesLocal(yamlSpecs)
+	swaggerSpecs := unmarshalManyFromYAML(editedYAMLSpecs, errs)
+
+	swaggerCommon, err := unmarshalFromYAML(yamlCommon)
+	if err != nil {
+		multierror.Append(errs, err)
```

Comment:
If parsing the common file fails, something is Borked. Since the common file is already treated as a special case, let's return an actual error (not a warning) if we fail to unmarshal common file